### PR TITLE
[react-color] Fixed typings and added missing properties

### PR DESCRIPTION
**Pull request by alexandercerutti**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:

https://github.com/casesandberg/react-color/blob/master/src/components/twitter/Twitter.js#L11 This line is common for all the components. A component accepts two properties that were not available. `className`  and `styles`. `styles` follows the rules of [`reactcss`](https://github.com/casesandberg/reactcss), from the same package creator. `react-color` uses `lodash/merge` to merge the properties coming in `styles`. For this reason, I've type'd `styles` following the format provided by `reactcss`. `reactcss` has been added as dependency in code import. I'm not sure if it needs a `package.json` to be included.

- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

It corrects the version number.

- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/react-color/lib/components/material/Material.d.ts
Comment (danger-public) [inline types/react-color/lib/components/material/Material.d.ts:6]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

**Comment by danger-public** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **types/react-color/lib/components/slider/Slider.d.ts#L6** - The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 70bfefd07e5b3b40fd493bc1859842dd23dd0515
</p>

File: types/react-color/lib/components/material/Material.d.ts
Comment (alexandercerutti) [reply types/react-color/lib/components/material/Material.d.ts:6]: I'm changing, but I haven't seen any .editorconfig style.

**Comment by alexandercerutti** [comment]
I'm applying also the change to the v2 folder, but what I'm noticing is that the package v3 is not out yet (stuck since 2018 I think?).
When the package is loaded, "v3" typings (typings in root) gets loaded. So I think the v2 folder should be removed and only one should be kept until the new major will come out.

**Comment by amcasey** [comment]
@alexandercerutti I'm not sure I understand your question about v2 and v3.  It sounds like you might be saying that the types somehow got ahead of the underlying library and there's no product code to go with the default (i.e. v3) types?  What did you mean about "inverting the types"?

**Comment by alexandercerutti** [comment]
@amcasey thanks for your reply.
For what I saw, if I go in VSCode and check the typings for one of the exported component, the index.d.ts used file is the one in the root, which is associated to v3, which hasn't been released yet.
You can see by yourself the version:
- https://www.npmjs.com/package/react-color
- https://github.com/casesandberg/react-color/tree/3.0.0

As you can see, the last commit on v3 was made 5 months ago and in production there's the v2.18.1.

So, since:
- I'm kinda new to this package and don't know why v2 folder has been created and v3 have been made at the top
- I actually don't how does folder version work in terms of Typescript declaration (never seen them anywhere, but when I ran the linting on typings with a lower version on top of the root index.d.ts, it gave me error)
- and on VSCode it seems like it is targeting not v2 typings but root typings (v3)

I was thinking about a solution, andI was thinking to create a v3 folder and put in the root the v2 files, but I don't know if it would ever work.

I hope you understand what I mean.
Let me know 😄

P.s. I didn't know which terms to use to search for the version splitting in typings. If you have any resources to link about it, it would be very appreciated 😊

**Comment by amcasey** [comment]
@alexandercerutti I think I understand your question, but I'll have to get back to you about the answer.

**Comment by amcasey** [comment]
@alexandercerutti, I'm not thrilled with the answer, but here's what I've got:
1) Yes, the react-color types are a mess.  Probably, the v2 types should go back in the root folder until v3 is sorted out, but that's a pretty big hammer and (rather surprisingly) we haven't been hearing too many complaints.
2) No, we don't presently support having a folder that's for a newer version than the root.  It could be handled, but this scenario is quite rare, so we're sticking to one-off solutions for now.
3) The mitigation for your editor experience is to specifically (dev) depend on a 2.x version of `@types/react-color`.  I'm unhappy that you're paying for someone else's mistake, but at least you'll get the right types.

I still owe you a link.  I'm not certain there is one.

**Comment by ErickTamayo** [comment]
I remember we had this discussion when I was fixing the definitions for this library. you can find here what we did about it: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35844#issuecomment-498895091

**Comment by amcasey** [comment]
Your memory is much sharper than mine. :smile:

I'm a little lost though - it looks like it was lamentable but acceptable then and the same is true now?  If you're asking about my point (1), what I meant was that a principled approach might discard the v3 types entirely, thereby restoring v2 to the root folder, but that we'd prefer to avoid taking such drastic action right now (not that your previous change had been incorrect).  However, I may have missed your point entirely or been correct either now or last June.

**Comment by alexandercerutti** [comment]
> Yes, the react-color types are a mess. Probably, the v2 types should go back in the root folder until v3 is sorted out, but that's a pretty big hammer and (rather surprisingly) we haven't been hearing too many complaints.

Maybe because most of the people still use this library in Javascript projects, so they won't need probably typings, or maybe they use a component that works correctly without changes for Typescript? I don't know, I'm supposing.

> No, we don't presently support having a folder that's for a newer version than the root. It could be handled, but this scenario is quite rare, so we're sticking to one-off solutions for now.

I don't understand what you mean exactly...

> The mitigation for your editor experience is to specifically (dev) depend on a 2.x version of @types/react-color. I'm unhappy that you're paying for someone else's mistake, but at least you'll get the right types.

Well, I actually looked the history for this package before sending this PR, and now, by reading the link you passed me, I'm understanding why this has been done.
One thing I'm sure is that the development of 3.0 and 2.0 proceeded in a parallel way, but 2.0 changed never got merged in 3.0, so also the typings kinda diverged (in fact, the current implementation, has some features I added to 3.0 that are available in 2.0 but not available in 3.0 yet).

And what I'm noticing is that, the last beta of `react-color` has been released on NPM like... [3 years ago](https://www.npmjs.com/package/react-color?activeTab=versions)!!

As per [the link you passed me](https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library), there's one thing I'm noticing, is that `tsconfig.json` lacks of `paths`, like it is happening with the `history` package.

```json
{
    "compilerOptions": {
        "baseUrl": "../../",
        "typeRoots": ["../../"],
        "paths": {
            "history": [ "history/v2" ]
        },
    },
    "files": [
        "index.d.ts",
        "history-tests.ts"
    ]
}
```

Could this be the reason my project is targeting v3 typings instead of v2 even if I installed the latest v2 that got published on NPM? I maybe should try to test the typings with that property in my project.

Also, I just saw that other react-color PRs has been accepted recently, after I opened my PR, so I should check what have been changed and apply them to my PRs.

**Comment by amcasey** [comment]
> Maybe because most of the people still use this library in Javascript projects, so they won't need probably typings, or maybe they use a component that works correctly without changes for Typescript? I don't know, I'm supposing.

If they are editing JS in VS or VS Code, I believe Automatic Type Acquisition will pull in the latest (i.e. v3) types.  If they've diverged from the v2 library that people are using, I'm surprised we're not hearing more complaints.

> > No, we don't presently support having a folder that's for a newer version than the root. It could be handled, but this scenario is quite rare, so we're sticking to one-off solutions for now.

> I don't understand what you mean exactly...

I understood your original question about "inverting" the types to be a question about whether v2 could go in the root folder and v3 could go in a subfolder.  With this structure, there would be "a folder that's for a newer version than the root".  We could add support for doing things that way, but it's easier to reason about the structure and detect mistakes if we keep the current constraint that the root folder must be for a newer version than any of the explicitly versioned folders.  Since we don't think many projects are in the same state as react-color (with an extremely stale beta), we haven't added that support.

>As per the link you passed me, there's one thing I'm noticing, is that tsconfig.json lacks of paths, like it is happening with the history package.

I'm not sure how this could cause the behavior you're seeing - I'll have to try it.  Maybe not having the path redirections is causing one of your dependencies to pull in v3 indirectly?

**Comment by amcasey** [comment]
>About paths, I looked better at what is the structure of the download @types/react-color in my current project. I currently have the v3.0.1 and I don't see the folder v2

The root (i.e. v3) folder is published as @types/react-color@3.x.x and does not contain a v2 folder.  The v2 folder is published as @types/react-color@2.x.x.  I don't see an update to that folder in the last five months, so there's been no republication in that time.

> But what's weird is that when I did npm install -D @types/react-color, it installed me the 3.0.1 and not the latest in terms of release date, so the 2.17.3 (I installed react-color two days before this PR).

If you're using TS 2.8 or TS 2.9, the newest compatible version of @types/react-color is 3.0.1.  To use 3.0.2, you need to be on at least TS 3.0.  (See the table at the top of the [npm page](https://www.npmjs.com/package/@types/react-color?activeTab=versions).)

> Can the DTBot publish two versions of the same package at the same time (for the same PR)?

If both folders are updated (e.g. because v2 has just been moved into a folder and types for v3 have been added), the publisher will publish both.

> Also, I actually don't even know if using paths in tsconfig.json would make the next version to have also v2 folder...

As far as I know, not updating the path maps is most likely to cause problems within the DT repo itself (e.g. because the v2 tests try to run against the v3 types).  I don't know of a way it could affect the npm packages.

**Comment by amcasey** [comment]
I'm told you'll need a new commit (or a rebase) to trigger the new CI mechanism and clear the `Where is GH Actions?` tag.

**Comment by alexandercerutti** [comment]
I'm always using the latest Typescript version available or the previous one. I'm on 3.9 right now.
Anyway, so no, I don't think anymore it is an issue of the `path`.
I mean, I was thinking that the package structure was equal to the repo one. I didn't know that different folders were being published as different versions of package. I was expecting as-is publishing.

So, maybe about this, there are actually no problems (one could have installed the correct version of the package typings - the problem is that one **must know** that). But yes, what's weird is that you did not receive many issues about the fact that v3 types (which is very stale).

___

So, here's a recap:

- I was thinking the package folder structure was reflecting the repo folder structure;
- By seeing that types were mismatching, I went to change types of v3 by cloning DT and I created this PR with changes to v3 that were actually already available (in a different - maybe worse? I don't know?) in the v2 folder. Therefore I brought both almost at the same level, probably including some changes that were available in the v2 to v3 even if they are not available (yet?) in the beta package - because the v3 is stale for a long time.
- It seems like not many people are experiencing what I experienced, probably because, in terms of interfaces, v2 and v3 do not differ very much (for example, v3 lacks a `styles` property to inject CSS on components that probably has been added after v3 branch creation). Since [API documentation](https://casesandberg.github.io/react-color/) is not really clear and well documented on the website or not updated, probably most of the usages follow what is explained on the website (and this is available in both v2 and v3). For example, I had to go to see the code to understand that I could have passed both `className` and `styles` (these attributes are not available in v3 - actually `className` was available only for one of the components, probably someone needed it and added it in the typings only to that component definition, but the feature is available to all the components in the package).

The fact, now, is this: if one does `npm install -D @types/react-color`, it gets the latest version, the 3.0.2 and not the latest "production" one. On the next `@types/react-color` PR, the problem will be the same. Also, accepting my PR, now, would mean that both packages are going to be updated, so the situation won't change.
So, even if we move v2 in the root and v3 in its own folder, I don't think this problem will never get solved.

Also, the [latest PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44685) for v3.0.2 of typings that happened 5 days ago, applied a change to v3 typings, as I mistakenly did, even if this change should have been made on v2.

So, what to do now? I'll have for sure to integrate the changes he made on both v2 and v3 to not break the things.

___

Is `Where is GH Actions?` tag an issue?

**Comment by amcasey** [comment]
I agree that the current state of affairs is confusing and that it's hard for users to recognize that they need a workaround and figure out how to apply it.  Our options would seem to be

1) Have `react-color` move v3 out of beta and ship.
2) Delete the v3 types and revert to using the v2 types by default.
3) Introduce a new DefinitelyTyped mechanism that allows v2 to be the default, even though v3 is available.
4) Merge this PR and live with the confusion.

(1) is beyond our control (though you might try following up on their repo).  (2) and (3) are possible without support from the react-color team, but would impact existing consumers of the types (some positively, some negatively) and I'm reluctant to recommend those options without hearing more complaints.  So my inclination would be to merge this PR to unblock people who need these APIs.

We recently changed our CI system and the `Where is GH Actions?` tag indicates that there are no results from the new CI system.  The easiest way to trigger a new run would be to add a new commit or rebase.

**Comment by alexandercerutti** [comment]
I agree with the possibilities. But PRs for this package should be checked better now that we know that this package has these problems.
A new DT mechanism would be interesting IMHO, but yes, maybe it would require a more detailed discussion. Where should it take place?

About the PR, I'll reflect as soon as possible, the changes of the last PR.
Anyway, for future PR, I might get notified. To allow the bot to tag me is it enough to add myself to the list on the top of `index.d.ts`?

**Comment by alexandercerutti** [comment]
So, @amcasey, what should we do about this? Are there any updates? Thank you. Anyway, I'm going to see the changes that the last PR made so I won't delete them (just in case).

**Comment by amcasey** [comment]
@alexandercerutti Sorry, I completely missed your response.

> But PRs for this package should be checked better now that we know that this package has these problems.
I can notify the other maintainers, but I'm not aware of a general mechanism for flagging this repo as special or providing an automatic reminder.

> Where should [further discussion] take place?
You can file a bug on this repo and tag @sandersn 

> About the PR, I'll reflect as soon as possible, the changes of the last PR.
You probably did this forever ago.  Sorry again.

> To allow the bot to tag me is it enough to add myself to the list on the top of index.d.ts?
Yes

**Comment by alexandercerutti** [comment]
@amcasey Not a problem at all. I just did them because I was busy creating other awesome things in typescript. 😁
Thank you for your reply.

I also already added myself to the list of contributors (for the root/v3 `index.d.ts`, but not for v2 - is it needed?).

About the bug report, I'll do that as soon as possible to discuss more this. Thanks.

**Comment by amcasey** [comment]
@alexandercerutti The latest version seems to rename a bunch of exported interfaces, but it looks like they were never really intended to be exported and aren't imported anywhere?

**Comment by amcasey** [comment]
Confirmed that adding yourself to the root `index.d.ts` should suffice.

**Comment by alexandercerutti** [comment]
> @alexandercerutti The latest version seems to rename a bunch of exported interfaces, but it looks like they were never really intended to be exported and aren't imported anywhere?

Hmm, I actually don't know. Which interfaces are you talking about?

**Comment by amcasey** [comment]
Looks like `ChromePickerDefaultStyles` became `ChromePickerStylesProps`, for example.

**Comment by amcasey** [comment]
I can see how extra usage of `CSSProperties` could lead to slower quick info.  The usage seems appropriate though.

**Comment by alexandercerutti** [comment]
@amcasey, yeah, probably those interfaces should not be exported.
By looking at the diffs, I see that ChromePicker's interfaces (and also the others maybe) were already being exported, so I continued on that way (some were exported types). For what concerns the names, I probably (I actually don't remember) tried to flat the differences between v2 and v3.

Anyway, I can remove them, I don't know if they might be useful outside (I don't think so, actually). Should I? Actually I don't even know if I should act on v3 anymore since it is stale.

Anyway, yes, `CSSProperties` is correct even if, to me, it is a shi**y (and poor-performing) way to handle CSS (outside Typescript)...

**Comment by alexandercerutti** [comment]
I noticed that both v2 and v3 components, have a default export like:

```typescript
export default class TwitterPicker extends Component<TwitterPickerProps> { }
```

I was thinking: should they be like this or is it the same?

```typescript
declare class TwitterPicker extends Component<TwitterPickerProps> {}
export default TwitterPicker;
```

**Comment by amcasey** [comment]
I'm not particularly worried about having extra imports, especially if they were already there.  I'm more concerned about removing (or renaming) things that were already exported.  However, in these particular cases, since the exports seems to have be unnecessary to begin with, the renames are probably fine.

You should merge your v3 changes if they match the state of the underlying library.  Otherwise, you can just leave v3 alone and make the v2 updates.

I'm now aware of a difference between `export default class` and `declare class; export default`.

**Comment by alexandercerutti** [comment]
@amcasey 

> However, in these particular cases, since the exports seems to have be unnecessary to begin with, the renames are probably fine.

Actually only "PickerProps" are getting exported, not the styles one. So I could remove those specific exports on both versions.

> You should merge your v3 changes if they match the state of the underlying library. Otherwise, you can just leave v3 alone and make the v2 updates.

I'll do a better comparison of this.

> I'm now aware of a difference between export default class and declare class; export default.

So... what? Should they be changed? Sorry, I did not understand 😅

**Comment by amcasey** [comment]
Oops, "now" should have been "not". 😅 No, I don't think you need to change your default exports.

**Comment by amcasey** [comment]
@alexandercerutti, my current understanding is that you're reviewing your v3 changes and this isn't ready yet.  Please let me know if I'm mistaken.

**Review by amcasey — APPROVED** [review]

**Comment by alexandercerutti** [comment]
Ready to merge

**Comment by alexandercerutti** [comment]
@amcasey I was noticing that [`@types/react-color`](https://www.npmjs.com/package/@types/react-color) has been published without a Readme... did I do something wrong perhaps?

**Comment by sandersn** [comment]
@alexandercerutti It looks like it's there to me: 

- https://unpkg.com/browse/@types/react-color@2.17.4/README.md
- https://unpkg.com/browse/@types/react-color@3.0.3/README.md

**Comment by alexandercerutti** [comment]
@sandersn but do you see it on NPM? I don't...

**Comment by sandersn** [comment]
Huh, you're right. It's there on an install as well, so it seems like (1) an error on NPM's page generation or (2) a change in NPM's page generation that now misses the README.

There's [an old issue about npm not *publishing* non-markdown READMEs on old versions of npm](https://github.com/npm/www/issues/136), but that exact bug isn't what's happening here.

**Pull request merged**
